### PR TITLE
#10163: Add backward support for fmod op

### PR DESCRIPTION
**Pull request by mouliraj-mcw**
Issue: [10163](https://github.com/tenstorrent/tt-metal/issues/10163)
### Binary Fmod_bw

**Test Files**

Unary Backward Fmod: `tests/ttnn/unit_tests/operations/backward/test_backward_fmod.py::test_bw_unary_fmod` - PASSED
Binary Backward Fmod : `tests/ttnn/unit_tests/operations/backward/test_backward_fmod.py::test_bw_binary_fmod` - PASSED

**Tested Input Range**

- Input Tensor: [-1e5 , 1e5]
- Other Tensor: [-1e5, 1e5]
- Grad Tensor: [-100, 100]

**Implemented formula :** ( - grad ) * trunc( input / other )

**Issue Faced**

Since it is a Div dependent op, the outputs of certain inputs are not computed correctly.
- Eg: For div(48, 48, rounding_mode("trunc") , expected result is 1 but the TT result is 0.
- Eg: For div(48, 24, rounding_mode("trunc") , expected result is 2 but the TT result is 1.
This creates a substantial difference in the final output.

**Steps to Resolve**
Added fp32 support to recip.

**Results**
Screenshot 2024-09-05 at 15 29 40

**CI :** [Link](https://github.com/tenstorrent/tt-metal/actions/runs/10718120597) - PASSED

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/binary_backward/device/binary_backward_op.cpp [hunk 1]
```diff
@@ -271,6 +271,15 @@ std::vector<ttnn::Tensor> _squared_difference_bw(
     return grad_tensor;
 }
 
+std::vector<Tensor> _fmod_bw(
+    const Tensor& grad, const Tensor& input, const Tensor& other, const MemoryConfig& output_mem_config) {
+    std::vector<Tensor> grad_tensor;
+    grad_tensor.emplace_back(grad);
+    Tensor result_div = div(input, other, true, "trunc");
+    Tensor grad_b = ttnn::multiply(ttnn::neg(grad), result_div, std::nullopt, output_mem_config);
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/eltwise/binary_backward/device/binary_backward_op.cpp:279]: Would you sanity check this?   I may have this totally wrong but should it be instead?
`Tensor grad_b = ttnn::multiply(ttnn::neg(result_div), grad, std::nullopt, output_mem_config);` ?

Comment (mouliraj-mcw) [reply ttnn/cpp/ttnn/operations/eltwise/binary_backward/device/binary_backward_op.cpp:279]: I tried using the suggested code, but it didn't resolve the issue. The problem stems from the div operation, which suffers from precision errors. The results from both methods are the same, as shown in [Results.pdf](https://github.com/user-attachments/files/16025931/Results.pdf)

**Comment by eyonland** [comment]
@rtawfik01 / @ttmtrajkovic , any thoughts on how we should deal with the precision issue from the div?

File: tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_fmod.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+import pytest
+import ttnn
+from tests.tt_eager.python_api_testing.unit_testing.backward_ops.utility_funcs import data_gen_with_range, compare_pcc
```
Comment (VirdhatchaniKN) [inline tests/tt_eager/python_api_testing/unit_testing/backward_ops/test_backward_fmod.py:8]: Same as [Link](https://github.com/tenstorrent/tt-metal/pull/9712/files#r1675359834)

File: ttnn/ttnn/__init__.py [hunk 3]
```diff
@@ -466,6 +466,7 @@ def manage_config(name, value):
     logaddexp2_bw,
     squared_difference_bw,
     add_bw,
+    fmod_bw,
```
Comment (VirdhatchaniKN) [inline ttnn/ttnn/__init__.py:469]: Same as [this](https://github.com/tenstorrent/tt-metal/pull/9712/files#r1675360757)

File: ttnn/ttnn/operations/binary_backward.py [hunk 4]
```diff
@@ -119,6 +119,12 @@ def _golden_function_backward_with_float(torch_op, grad_tensor, input_tensor_a,
     )
 )(ttnn._ttnn.operations.binary_backward.squared_difference_bw)
 
+fmod_bw = ttnn.register_operation(
```
Comment (VirdhatchaniKN) [inline ttnn/ttnn/operations/binary_backward.py:122]: Same as [this](https://github.com/tenstorrent/tt-metal/pull/9712/files#r1675360757)

**Review by Aswinmcw — APPROVED** [review]
LGTM

**Review by eyonland — APPROVED** [review]

**Pull request merged**